Create full CRD structure

dkarpele opened the pull request:
Important highlights:
1. `metadata.namespace` is a namespace for ImageUpdater CR.
2. 'spec.namespace' is a namespace where the controller will look for Argo CD Applications matching the criteria in `ApplicationRefs`. 'spec.namespace' **is required**. It means that all Applications in each CR will be assigned to one and only one namespace. 
3. `ApplicationRefs`, `images` are required. 

Full CRD version is `config/crd/bases/argocd-image-updater.argoproj.io_imageupdaters.yaml`

Short version:

```yaml
apiVersion: argocd-image-updater.argoproj.io/v1alpha1 
kind: ImageUpdater 
metadata:
  name: "imageupdater-sample"
  namespace: "argocd"
spec:
  namespace: "argocd"
  updateStrategy: "semver" # Default strategy if not specified per image
  forceUpdate: true
  allowTags: "regexp:^[0-9a-f]{7}$"
  ignoreTags: ["latest", "master"]
  pullSecret: "secret:<namespace>/<secret_name>#<field>" 
  writeBackMethod: "git:secret:argocd-image-updater/git-creds" # e.g., "git", "argocd" (to update Application CR). Default argocd - optional.
  gitConfig: # Group git-related settings. Should be ignored when writeBackMethod: "argocd"
    repository: "git@github.com:example/example.git"
    branch: "main"
    writeBackTarget: "helmvalues:/helm/config/test-values.yaml" # OR "kustomization:/config/overlays/bar" OR "helmvalues:/{{ .app.path.path }}/values.yaml" - for ApplicationSet.
  applicationRefs: # It's a list of applications or 1 app.
    - namePattern: "name1*"
      ...
    - namePattern: "name2*"
    ...
    - namePattern: "name3*"
      labelSelectors:
        matchLabels:
          app: "true"

      updateStrategy: "semver" # strategy for name3*
      forceUpdate: true
      allowTags: "regexp:^[0-9a-f]{7}$"
      ignoreTags: ["latest", "master"]
      pullSecret: "secret:<namespace>/<secret_name>#<field>" 
      writeBackMethod: "git:secret:argocd-image-updater/git-creds" # e.g., "git", "argocd" (to update Application CR). Default argocd - optional.
      gitConfig: # Group git-related settings. Should be ignored when writeBackMethod: "argocd"
        repository: "git@github.com:example/example.git"
        branch: "main"
        writeBackTarget: "helmvalues:/helm/config/test-values.yaml" # OR "kustomization:/config/overlays/bar" OR "helmvalues:/{{ .app.path.path }}/values.yaml" - for ApplicationSet.

  # Image List for the App and Per-Image Overrides
  # (Maps to argocd-image-updater.argoproj.io/image-list and <alias>.* annotations)
  images:
  - alias: "nginx" # Maps to current '<alias>=' part of image-list annotation. Should clearly separate the image name (for registry) from its initial version/tag. What we are currently using before = sign (image=image:version)
    imageName: "nginx:1.17.10" # The actual image name for registry lookup OR 10.42.0.1:30000/test-image:latest
    updateStrategy: "semver" # Per-image override
    forceUpdate: true
    allowTags: "regexp:^[0-9a-f]{7}$"
    ignoreTags: ["latest", "master"]
    pullSecret: "secret:<namespace>/<secret_name>#<field>"
    platforms: ["linux/arm64", "linux/amd64"] 
    manifestTargets:
      helm:
        name: "dex.image.name" 
        tag: "dex.image.tag" # e.g., for nginx.image.repository
        spec: "image/name:1.0"
      kustomize:
        name: "quay.io/argoproj/argocd"
status:
  lastUpdatedAt: <timestamp>
  imageStatus:
    - name: nginx
      version: "1.21"
      applications:
        - appName: ns1/example1
          lastUpdatedAt: <timestamp>
        - ns2/example2
          lastUpdatedAt: <timestamp>
  conditions: 
```

chengfang commented on api/v1alpha1/imageupdater_types.go:
```diff
@@ -24,81 +24,243 @@ import (
+	// This acts as the default if not overridden at a more specific level.
+	// +optional
+	// +kubebuilder:default:="semver"
+	UpdateStrategy *string `json:"updateStrategy,omitempty"`
```
UpdateStrategy-related fields appear in all 3 levels. Does it make sense to define a common struct and embed it into the 3 structs, so that these fields can directly be referenced as if they were directly defined in these 3 levels?

dkarpele replied:
Currently Update Strategy can be defined on 2 levels:
1. Application-wide `argocd-image-updater.argoproj.io/update-strategy: digest`
In this imageupdater_types.go it's an `ApplicationRef.UpdateStrategy`
2. Per-image `argocd-image-updater.argoproj.io/myimage.update-strategy: digest`
In this imageupdater_types.go it's an `ImageConfig.UpdateStrategy`

I have decided to define `ImageUpdaterSpec.UpdateStrategy` globally. Thanks to this field user can define update strategy only once instead of duplicating it per application. 

Example shows that user has 2 applications with the same updateStrategy semver.
```
spec:
  namespace: "argocd"
  updateStrategy: "semver" # Default strategy if not specified per image
  applicationRefs: # It's a list of applications or 1 app.
    - namePattern: "name1*"
    - namePattern: "name2*"
```

Example shows that user has 2 applications with the same updateStrategy semver. But the third app has startegy digest that will have a priority.
```
spec:
  namespace: "argocd"
  updateStrategy: "semver" # Default strategy if not specified per image
  applicationRefs: # It's a list of applications or 1 app.
    - namePattern: "name1*"
    - namePattern: "name2*"
    - namePattern: "name3*"
      updateStrategy: "digest" 
```

Example shows that user has 2 applications with the same updateStrategy semver. But the third app has startegy digest that will have a priority. Image has its own strategy strategy latest that will have a priority. 
```
spec:
  namespace: "argocd"
  updateStrategy: "semver" # Default strategy if not specified per image
  applicationRefs: # It's a list of applications or 1 app.
    - namePattern: "name1*"
    - namePattern: "name2*"
    - namePattern: "name3*"
      updateStrategy: "digest" 
  images:
  - alias: "nginx" # Maps to current '<alias>=' part of image-list annotation. Should clearly separate the image name (for registry) from its initial version/tag. What we are currently using before = sign (image=image:version)
    imageName: "nginx:1.17.10" # The actual image name for registry lookup OR 10.42.0.1:30000/test-image:latest
    updateStrategy: "latest" # Per-image override
```

chengfang replied:
> UpdateStrategy-related fields appear in all 3 levels. Does it make sense to define a common struct and embed it into the 3 structs, so that these fields can directly be referenced as if they were directly defined in these 3 levels?

If we omit the field name (e.g., `GlobalCommonUpdateSettings`), we can *directly* access all fields in `CommonUpdateSettings` from a `ImageUpdaterSpec`, for ex:

```
	imageUpdaterSpec := ImageUpdaterSpec{
		CommonUpdateSettings: &CommonUpdateSettings{
			UpdateStrategy: "argocd",
			ForceUpdate:    false,
			IgnoreTags:     []string{"xxx"},
			PullSecret:     "xxx",
		},
	}
	imageUpdaterSpec.ForceUpdate = true  // ,or longer form next line
	imageUpdaterSpec.CommonUpdateSettings.ForceUpdate = true
	
	imageUpdaterSpec.UpdateStrategy = "git"  // or longer form next line
	imageUpdaterSpec.CommonUpdateSettings.UpdateStrategy = "git"
```

dkarpele replied:
Thank you for the review! I updated this.

chengfang commented on api/v1alpha1/imageupdater_types.go:
```diff
@@ -24,81 +24,243 @@ import (
+	// +optional
+	AllowTags *string `json:"allowTags,omitempty"`
+
+	// IgnoreTags is a list of tags to ignore.
```
From https://argocd-image-updater.readthedocs.io/en/latest/configuration/images/#appendix , ignoreTags is a list of glob-like patterns of tags.

dkarpele replied:
I can change the comment to something like: `// IgnoreTags is a list of of glob-like patterns of tags to ignore.`

chengfang replied:
I'm also a bit uncomfortable with the mixed use of string vs *string, bool vs *bool in struct fields. I think the common practice in this project is to use values especially for primitives in struct fields.

dkarpele replied:
Right, in argocd types.go values are used for primitives https://github.com/argoproj/argo-cd/blob/release-2.13/pkg/apis/application/v1alpha1/types.go

dkarpele commented:
I added an example into config/samples. 
Also added 
`// +kubebuilder:validation:XValidation:rule="has(self.helm) ? !has(self.kustomize) : has(self.kustomize)",message="Exactly one of helm or kustomize must be specified within manifestTargets if the block is present."`

to validate that only helm or kustomize can be used under ManifestTarget
